### PR TITLE
Initial script

### PR DESCRIPTION
**Pull request by mahyar-osn**
This is the very initial script to fetch data from MoH, pre-process, and create plots.
Extensions and modifications will be done as required by the OpenCOR and the modeling pipeline specifics.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+run_data.plot_daily_trend()
+run_data.plot_cumulative_sum()
+
+plt.show()
+
+```
+
+Running the above code in a file should produce the following two figures:
+
+![alt text](resources/Figure_1.png)
```
Comment (agarny) [inline README.md:32]: I would probably update the legend to read something like "Daily confirmed cases", "Daily probable cases" and "Daily cases".

File: README.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+![alt text](resources/Figure_2.png)
```
Comment (agarny) [inline README.md]: The legend for Figure 2 seems wrong to me. Shouldn't it be something like "Total confirmed cases", "Total probable cases" and "Grand total"?

File: README.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+```python
+from src.moh_data.main import Basic
+import matplotlib.pyplot as plt
+
+run_data = Basic()
+run_data.plot_daily_trend()
+run_data.plot_cumulative_sum()
+
+plt.show()
+
```
Comment (agarny) [inline README.md]: Delete unnecessary empty line.

File: setup.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
+from setuptools import setup, find_packages
```
Comment (agarny) [inline setup.py:1]: If you use PyCharm CE, then I would recommend using the `Code` | `Reformat Code` menu item.

File: setup.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    version="0.0.1",
+    description="Simple script to automatically fetch, process, and plot the NZ covid-19 cases data",
+    long_description="\n".join(readme) + source_license,
+    classifiers=[
+        "Programming Language :: Python",
+        "Topic :: Scientific/Engineering :: Medical Science Apps."
+    ],
+    author="Auckland Bioengineering Institute",
+    author_email="m.osanlouy@auckland.ac.nz",
+    url="https://github.com/mahyar-osn/covid19_nz_data_processing/",
```
Comment (agarny) [inline setup.py]: Need to update the URL.

File: setup_exe.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+if sys.platform == 'win32':
+    base = 'Win32GUI'
+
+executables = [
+    Executable('src/moh-data/app/app.py', base=base)
+]
+
+
+setup(
+    name="covid19_nz_data_processing",
```
Comment (agarny) [inline setup_exe.py]: Need updating.

File: setup_exe.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    version="0.0.1",
+    description="Simple script to automatically fetch, process, and plot the NZ covid-19 cases data",
+    executables=executables
+)
```
Comment (agarny) [inline setup_exe.py]: Need trailing empty line.

File: setup_exe.py [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+executables = [
+    Executable('src/moh-data/app/app.py', base=base)
+]
+
+
+setup(
+    name="covid19_nz_data_processing",
+    version="0.0.1",
+    description="Simple script to automatically fetch, process, and plot the NZ covid-19 cases data",
```
Comment (agarny) [inline setup_exe.py]: Rephrase: "Script to automatically fetch, process, and plot the NZ Covid-19 cases data."

File: setup.py (hunk 4 above)
Comment (agarny) [reply setup.py:1]: Same with other Python files in the repo.

File: src/moh_data/app/app.py [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+
+def main():
+    app = QtGui.QApplication(sys.argv)
+    app.setStyle(QtGui.QStyleFactory.create("Cleanlooks"))
+    app.setPalette(QtGui.QApplication.style().standardPalette())
+    ui = PlotWidget()
+    sys.exit(app.exec_())
+
+
+# if __name__ == '__main__':
```
Comment (agarny) [inline src/moh_data/app/app.py]: Delete if not needed.

File: src/moh_data/app/app.py [hunk 10]
```diff
@@ -0,0 +1,18 @@
+
```
Comment (agarny) [inline src/moh_data/app/app.py]: Delete unnecessary empty line.

**Review by agarny — COMMENTED** [review]
- A few minor issues reported as comments.
- Delete empty files, i.e. `src/__init__.py`, `src/moh_data/__init__.py`, `src/moh_data/app/__init__.py`, and `src/moh_data/core/__init__.py`.
- I haven't checked the PyQt code as such, but is that code needed? I thought the idea of this script was to retrieve the data from the MoH. If you want to plot things then it might be good enough to use matplotlib?
- Is your script only for Windows? I mean, what about `src/moh_data/app/tcl86t.dll` and `src/moh_data/app/tk86t.dll`?

**Comment by mahyar-osn** [comment]
> * A few minor issues reported as comments.

Made the changes. One question. Python PEP8 requires one empty line at the end of a script. I see that you mentioned that I should remove them. I do not have any preference either way, just checking.

> * Delete empty files, i.e. `src/__init__.py`, `src/moh_data/__init__.py`, `src/moh_data/app/__init__.py`, and `src/moh_data/core/__init__.py`.

Done. Always forget that Py36 doesn't need them anymore.... 

> * I haven't checked the PyQt code as such, but is that code needed? I thought the idea of this script was to retrieve the data from the MoH. If you want to plot things then it might be good enough to use matplotlib?

Initially the idea was to retrieve data and make some plots and generate tables. The PyQt side of the thing was just a toy app to plot these figures (was done for P.H. for building a simple exe file)

> * Is your script only for Windows? I mean, what about `src/moh_data/app/tcl86t.dll` and `src/moh_data/app/tk86t.dll`?

It should be compatible with any platform. Those dll files I only provided them if someone wanted to build an exe file in Windows. 

I am happy to get rid of the /app/ directory as a whole, since we would only want to provide a data streaming.

**Comment by agarny** [comment]
> Made the changes. One question. Python PEP8 requires one empty line at the end of a script. I see that you mentioned that I should remove them. I do not have any preference either way, just checking.

I agree with one trailing empty line and that's what I do (and what GitHub also likes to see). So, not quite sure which occurrence you are talking about.

> I am happy to get rid of the /app/ directory as a whole, since we would only want to provide a data streaming.

Yes, I think I would do that.

**Comment by mahyar-osn** [comment]
> I agree with one trailing empty line and that's what I do (and what GitHub also likes to see). So, not quite sure which occurrence you are talking about.

Actually never mind, I got the ReadMe trailing line confused.

> Yes, I think I would do that.

Sure, will do it in the next PR. Need to make some changes.

**Comment by agarny** [comment]
> Sure, will do it in the next PR. Need to make some changes.

?? Why not this PR?

**Comment by mahyar-osn** [comment]
> ?? Why not this PR?

I guess good point :)

**Review by agarny — APPROVED** [review]
The code is not as clean as I would like it to be (i.e. use PyCharm to reformat the code, remove unneeded comments/code, etc.), but that's the "perfectionist" in me.

So, as an initial script, I guess it's fine. However, as a potential consumer of that module, I cannot do much with it. Clearly, all the information I would need is there, but it's all buried in the module while I would expect a simple class that fetches the data from the MoH and makes it available to consumers through some simple methods.

**Comment by mahyar-osn** [comment]
In terms of code cleaning/reformatting, I always use the PyCharm reformatting and from what I can see the code looks OK i.e. no conflict with PEP8. Could you be more specific?

Additionally, the `main.py` contains four methods to get the data that get plotted e.g. `get_daily_data()`, `get_daily_cumulative_data()` etc. However, once I know exactly what kind of data is required for SEIR model, I can easily change the code to provide easy methods to fetch the desired data.

**Comment by agarny** [comment]
> In terms of code cleaning/reformatting, I always use the PyCharm reformatting and from what I can see the code looks OK i.e. no conflict with PEP8. Could you be more specific?

Sure. If you do a `Reformat Code` on the whole module you then get the following `git diff`:

```diff
diff --git a/setup.py b/setup.py
index f68cf2e..8f8c4f5 100644
--- a/setup.py
+++ b/setup.py
@@ -1,6 +1,7 @@
-from setuptools import setup, find_packages
-import os
 import io
+import os
+
+from setuptools import setup, find_packages

 SETUP_DIR = os.path.dirname(os.path.abspath(__file__))

diff --git a/src/moh_data/core/collector.py b/src/moh_data/core/collector.py
index c301c9e..52533f5 100644
--- a/src/moh_data/core/collector.py
+++ b/src/moh_data/core/collector.py
@@ -1,8 +1,8 @@
-from .query import FindExcelFile
-
 import numpy as np
 import pandas as pd

+from .query import FindExcelFile
+

 def _check_zero(data, cols):
     for col in cols:
diff --git a/src/moh_data/core/query.py b/src/moh_data/core/query.py
index 3653b4d..f330006 100644
--- a/src/moh_data/core/query.py
+++ b/src/moh_data/core/query.py
@@ -1,4 +1,5 @@
 from urllib.request import urlopen
+
 import requests
 from bs4 import BeautifulSoup, SoupStrainer

diff --git a/src/moh_data/utils/visualisation.py b/src/moh_data/utils/visualisation.py
index 587049e..2c770f1 100644
--- a/src/moh_data/utils/visualisation.py
+++ b/src/moh_data/utils/visualisation.py
@@ -51,4 +51,4 @@ class Visualisation:
         plt.show()

     def _get_last_available_date(self):
-        return self._df.index[self._df.index.argsort() == (len(self._df.index)-2)][0]
+        return self._df.index[self._df.index.argsort() == (len(self._df.index) - 2)][0]
```

> Additionally, the `main.py` contains four methods to get the data that get plotted e.g. `get_daily_data()`, `get_daily_cumulative_data()` etc. However, once I know exactly what kind of data is required for SEIR model, I can easily change the code to provide easy methods to fetch the desired data.

Yes, I saw those, but those are raw data. I wish there were some methods that would give me a number for a given date. For example, `confirmed_cases(day=<today>)`, `total_confirmed_cases(day=<today>)`, etc. i.e. return the number of x for the current day (default) or any day (if a day is passed as parameter). As I said, that information is available, just burried and therefore not easily/directly accessible.

**Comment by mahyar-osn** [comment]
> Sure. If you do a `Reformat Code` on the whole module you then get the following `git diff`:

Thanks Alan. Just did that and got the following message:

`D:\covid-19\codes\python_packages\moh_data (master)`
`(sparc-venv36) λ git diff`
`diff --git a/src/moh_data/utils/visualisation.py b/src/moh_data/utils/visualisation.py`
`index 587049e..2c770f1 100644`
`--- a/src/moh_data/utils/visualisation.py`
`+++ b/src/moh_data/utils/visualisation.py`
`@@ -51,4 +51,4 @@ class Visualisation:`
`         plt.show()`

`     def _get_last_available_date(self):`
`-        return self._df.index[self._df.index.argsort() == (len(self._df.index)-2)][0]`
`+        return self._df.index[self._df.index.argsort() == (len(self._df.index) - 2)][0]`

Will commit this change now.

> Yes, I saw those, but those are raw data. I wish there were some methods that would give me a number for a given date. For example, `confirmed_cases(day=<today>)`, `total_confirmed_cases(day=<today>)`, etc. i.e. return the number of x for the current day (default) or any day (if a day is passed as parameter). As I said, that information is available, just burried and therefore not easily/directly accessible.

I see. Sure, that can be done. Will start a new PR once I add these methods to the code. Might have to iterate with you on this to ensure all the data are available the way you would like them for the SEIR model.

**Comment by agarny** [comment]
> I see. Sure, that can be done. Will start a new PR once I add these methods to the code. Might have to iterate with you on this to ensure all the data are available the way you would like them for the SEIR model.

Sure, no problem. I need to do some CellML 2.0 stuff now, but will hopefully get back to my MoH data needs later today.

**Pull request merged**